Feature/rgb object detection

minhnh opened the pull request:
---
name: RGB object detection
about: add general architecture for object detection and related necessary features
---

## Description
* add `ImageDetector` abstract class for different implementations of image detection models
* add image detection action server which works with extensions of abstract class `ImageDetector`
* add test script which works with extensions of `ImageDetector` and takes image directory, image topic or point cloud topic as image source
* add boost conversion between `numpy.ndarray` and `cv::Mat` to allow easy extraction of images and point coordinates
* add point cloud utilities in C++ and corresponding Python interfaces
* add a definition for 2D bounding boxes in C++ with a Python interface for common visualization and point cloud processing methods

## Motivation and Context
Simplify trying out new image detection models and unify C++ and Python logics for image and point cloud processing.

## How Has This Been Tested?
RGB detection is tested on Lucy and locally with Xtion camera.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Functionalities included
- Robot independent

## Checklist:
- [x] I have updated the documentation accordingly.
- [x] I have updated the `package.xml` with the correct dependencies.
- [ ] I have updated the CHANGELOG.rst file
- [x] My code follows the code style of this project.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

minhnh committed:
add vis. functions and RGB detector base class

minhnh committed:
add detection service and detection test class

- test class returns random boxes
- add example kwargs file for test class
- add example class annotation file
- add BoundingBox2D message, ImageDetection.srv

minhnh committed:
add test script and launch file for detection

minhnh committed:
move loading kwargs file to base class

minhnh committed:
add boost python conversion for cv::Mat

- move draw bounding boxes to C++ to test cv::Mat conversion
- move generating detection message from prediction to separate method
- add DetectionServiceProxy sub class for image detection
- remove generate_msgs
- handle roslint errors

minhnh committed:
Merge branch 'refactor/move_perception_libs' of https://github.com/minhnh/mas_perception into feature/rgb_object_detection

minhnh committed:
replace enum with struct 2D for bounding boxes

minhnh committed:
expose fit_box_to_image and crop_image to python

- refactor cropImage to common/bounding_box_2d
- replace cv::Rect field with function
- add type check in visualization.py

minhnh committed:
add image detection from cloud and image topics

- add methods to extract image from organized clouds, expose them to Python
- add functionality to image_detection_client_test: detect objects from cloud/image topic
- add method ImageDetector.detection_msg_to_bounding_boxes()
- add visualization.draw_labeled_boxes_img_msg() to draw boxes on image messages

minhnh committed:
add cropping point cloud using 2D box

minhnh committed:
add extracting coordinates with bounding box

minhnh committed:
change launch namespace

minhnh committed:
add drawing labeled boxes for image messages

also move boost python related to _cpp_wrapper target for correct linking

minhnh committed:
add cloud transform python interface

minhnh committed:
update frame_id in transform_point_cloud()

minhnh committed:
Merge branch 'refactor/move_perception_libs' of https://github.com/minhnh/mas_perception into feature/rgb_object_detection

minhnh committed:
Merge branch 'refactor/move_perception_libs' of https://github.com/minhnh/mas_perception into feature/rgb_object_detection

minhnh committed:
refactor image recognition

minhnh committed:
add image detection object server

minhnh committed:
rewrite test script to share action server logic

minhnh committed:
remove messages added for detection service

minhnh committed:
Merge branch 'refactor/move_perception_libs' of https://github.com/minhnh/mas_perception into feature/rgb_object_detection

minhnh committed:
Merge branch 'refactor/move_perception_libs' of https://github.com/minhnh/mas_perception into feature/rgb_object_detection

minhnh committed:
refactor and add documentation

- refactor preprocessing image messages to base class for ImageDetector
- add comments and documentation

minhnh committed:
finalize documentation for C++ library

minhnh committed:
add in-code comments, restructure Python doc

minhnh committed:
Merge branch 'refactor/move_perception_libs' of https://github.com/minhnh/mas_perception into feature/rgb_object_detection

minhnh committed:
update Python documentation

minhnh committed:
remove unused detection test server

minhnh committed:
update README.md

minhnh committed:
Merge branch 'kinetic' of https://github.com/b-it-bots/mas_perception into feature/rgb_object_detection

argenos commented:
This should probably merged into *devel*

minhnh commented:
I didn't realize that this repo has a `devel` branch. Sure I can switch target.

minhnh committed:
add usage example

alex-mitrevski commented on mas_perception_libs/ros/src/mas_perception_libs/image_detector.py:
```diff
@@ -0,0 +1,242 @@
+    Y_MAX = 5
+
+
+class ImageDetector(object):
```
I'd rename this to `ImageDetectorBase` to indicate that it's an abstract class.

minhnh replied:
renamed in latest commit

alex-mitrevski commented on mas_perception_libs/README.md:
```diff
@@ -11,6 +11,37 @@ Contains shared perception defintions in C++.
 Contains shared perception definitions in Python.
```
Your documentation is generally great, so nicely done! I do have a couple of improvement suggestions though:
* since `perception_libs` is a relatively large piece of software, it would be very helpful to add a high-level diagram in the README introduction (or a set of diagrams - one per section) that gives an overview of the whole library and thus helps anyone who's trying to use it
* I think the introduction should also explain the rationale behind having both a C++ library and a Python interface (i.e. what's the purpose of both); short descriptions of the two should also be included in the sections dedicated to them

minhnh replied:
added image detection diagram and explanation for C++ Python interface

minhnh committed:
rename to ImageDetectorBase

sthoduka commented on mas_perception_libs/common/include/mas_perception_libs/bounding_box_2d.h:
```diff
@@ -0,0 +1,105 @@
+            : mX(pX), mY(pY), mWidth(pWidth), mHeight(pHeight), mLabel(std::move(pLabel)), mColor(pColor)
+    { }
+
+    BoundingBox2D(std::string pLabel, int pX, int pY, int pWidth, int pHeight)
```
I think passing arguments by const reference is best practice for non-primitive datatypes (i.e.  `const std::string &pLabel` and `const cv::Scalar &pColor`)

https://www.learncpp.com/cpp-tutorial/73-passing-arguments-by-reference/

Can you also look at all your other function arguments?

minhnh replied:
updated all unmodified params to const reference with the last commit.

sthoduka commented on mas_perception_libs/ros/src/mas_perception_libs/scene_detection_action.py:
```diff
@@ -24,15 +33,133 @@ def _execute_cb(self, goal):
+                                          BoundingBox2D((0, 0, cropped_cloud.width, cropped_cloud.height)))
+        mean_coord = np.nanmean(np.reshape(cropped_coord, (-1, 3)), axis=0)
+        min_coord = np.nanmin(np.reshape(cropped_coord, (-1, 3)), axis=0)
+        pose_stamped.pose.position.x = min_coord[0]
```
can you describe here how the pose will look for an object? i.e. which directions are y and z? 
Are these assumptions valid for different types of grasps (top grasp, side grasp etc.)

minhnh replied:
I removed this part for now since it overlaps with pose calculation in `ObjectDetector`. I will think on this a bit more and make a different PR.

minhnh committed:
refactor & update documentation

- change parameters to const ref as needed
- rename fields to match class name
- add intro to library in README

minhnh committed:
add image detection architecture

minhnh committed:
fix image link

minhnh committed:
remove pose estimation funtion

sthoduka commented on mas_perception_libs/common/include/mas_perception_libs/bounding_box_2d.h:
```diff
@@ -0,0 +1,105 @@
+ * @param pFontScale
+ */
+void
+drawLabeledBoxes(cv::Mat &pImage, std::vector<BoundingBox2D> pBoundingBoxes,
```
the std::vector should be const reference here, right? Or is it  being modified?

minhnh replied:
If the bounding boxes are modified, can the vectors be const ref?

minhnh replied:
update documentation about boxes will be adjusted to fit image size.

sthoduka commented on mas_perception_libs/common/include/mas_perception_libs/bounding_box_2d.h:
```diff
@@ -0,0 +1,105 @@
+ * @return cropped CV image
+ */
+cv::Mat
+cropImage(cv::Mat &pImage, std::vector<cv::Point2f> &pVertices, int pOffset = 0, bool pCopy = true);
```
same here

minhnh replied:
changed to const ref

sthoduka commented on mas_perception_libs/common/include/mas_perception_libs/bounding_box_2d.h:
```diff
@@ -0,0 +1,105 @@
+ * @brief crop a CV image using a BoundingBox2D object
+ */
+cv::Mat
+cropImage(cv::Mat &pImage, BoundingBox2D &pBox, int pOffset = 0, bool pCopy = true);
```
const ref: pBox

minhnh replied:
The bounding boxes are all adjusted to fit within image size before cropping/drawing is done. So they need to be mutable. I could also make the `fitToImageSize` function return a new object, but it seems reasonable to modify the object as a mutable parameter.

sthoduka commented on mas_perception_libs/ros/include/mas_perception_libs/image_bounding_box.h:
```diff
@@ -39,8 +40,13 @@ class ImageBoundingBox
+ *        bounding_box_2d.h in the background.
+ */
+sensor_msgs::ImagePtr
+drawLabeledBoxesImgMsg(const sensor_msgs::Image&, std::vector<BoundingBox2D>,
```
here are you modifying the sensor_msgs::Image? in that case const would be a problem. And the vector should probably be const ref.

minhnh replied:
Image will be copied before boxes are drawn, documentation has been added for the parameter about this.

sthoduka commented on mas_perception_libs/ros/include/mas_perception_libs/point_cloud_utils.h:
```diff
@@ -0,0 +1,53 @@
+ *        defined by a BoundingBox2D object
+ */
+cv::Mat
+cropCloudToXYZ(const PointCloud &pCloud, BoundingBox2D &pBox);
```
all these BoundingBox2D objects should be const ref too I think

minhnh replied:
similar reason to above

minhnh committed:
document modifying box size to fit cloud, images
